src/platform/ESP32/MdnsImpl.cpp

```diff
@@ -0,0 +1,241 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "lib/protocols/mdns/platform/Mdns.h"
+
+#include <esp_err.h>
+#include <lwip/ip4_addr.h>
+#include <lwip/ip6_addr.h>
+#include <mdns.h>
+
+#include "platform/CHIPDeviceLayer.h"
+#include "support/CHIPMem.h"
+#include "support/CodeUtils.h"
+#include "support/logging/CHIPLogging.h"
+
+namespace {
+
+static constexpr uint32_t kTimeoutMilli = 3000;
+static constexpr size_t kMaxResults     = 20;
+
+} // namespace
+
+namespace chip {
+namespace Protocols {
+namespace Mdns {
+
+CHIP_ERROR ChipMdnsInit(MdnsAsnycReturnCallback initCallback, MdnsAsnycReturnCallback errorCallback, void * context)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    esp_err_t espError;
+    uint16_t discriminator;
+    char hostname[11]; // Hostname willbe "CHIP-XXXX"
+
+    espError = mdns_init();
+    VerifyOrExit(espError == ESP_OK, error = CHIP_ERROR_INTERNAL);
+    SuccessOrExit(error = DeviceLayer::ConfigurationMgr().GetSetupDiscriminator(discriminator));
+    snprintf(hostname, sizeof(hostname), "CHIP-%04d", discriminator);
+    espError = mdns_hostname_set(hostname);
+    VerifyOrExit(espError == ESP_OK, error = CHIP_ERROR_INTERNAL);
+
+exit:
+    if (espError != ESP_OK)
+    {
+        ChipLogError(DeviceLayer, "esp mdns internal error: %s", esp_err_to_name(espError));
+    }
+    initCallback(context, error);
+
+    return error;
+}
+
+static const char * GetProtocolString(MdnsServiceProtocol protocol)
+{
+    return protocol == MdnsServiceProtocol::kMdnsProtocolTcp ? "_tcp" : "_udp";
+}
+
+CHIP_ERROR ChipMdnsPublishService(const MdnsService * service)
+{
+    CHIP_ERROR error        = CHIP_NO_ERROR;
+    mdns_txt_item_t * items = nullptr;
+    esp_err_t espError;
+
+    VerifyOrExit(service->mTextEntrySize <= UINT8_MAX, error = CHIP_ERROR_INVALID_ARGUMENT);
+    if (service->mTextEntryies)
+    {
+        items = static_cast<mdns_txt_item_t *>(chip::Platform::MemoryCalloc(service->mTextEntrySize, sizeof(mdns_txt_item_t)));
+        VerifyOrExit(items != nullptr, error = CHIP_ERROR_NO_MEMORY);
+        for (size_t i = 0; i < service->mTextEntrySize; i++)
+        {
+            items[i].key = service->mTextEntryies[i].mKey;
+            // Unfortunately ESP mdns stack dosen't support arbitrary binary data
+            items[i].value = reinterpret_cast<const char *>(service->mTextEntryies[i].mData);
+        }
+    }
+
+    espError = mdns_service_add(service->mName, service->mType, GetProtocolString(service->mProtocol), service->mPort, items,
+                                service->mTextEntrySize);
+    // The mdns_service_add will return error if we try to add an existing service
+    if (espError != ESP_OK)
```

Comment:
there are other errors, too (e.g. no memory).  perhaps verify that mdns_service_add() is complaining about an existing service, if possible?

another option is to fail registration as does the underlying API, and have that semantic bubble-up through the CHIP API fro Mdns.